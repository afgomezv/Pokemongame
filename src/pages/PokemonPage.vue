<template>
  <h1 v-if="!pokemon">Espere por favor...</h1>

  <div v-else>
    <h1>¿Quién es este pokémon?</h1>
    <!-- TODO: img -->
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <!-- TODO: Opciones -->
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />
  </div>
  <template v-if="showAnswer">
    <h2 class="fade-in">{{ message }}</h2>
    <button @click="newGame">Nuevo Juego</button>
  </template>
</template>

<script>
import PokemonOptions from "@/components/PokemonOptions.vue";
import PokemonPicture from "@/components/PokemonPicture.vue";

import getPokemonOptions from "@/helpers/getPokemonOptions";

//console.log(getPokemonOptions());

export default {
  components: {
    PokemonOptions,
    PokemonPicture,
  },

  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
    };
  },

  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      //console.log(this.pokemonArr);
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(selectId) {
      this.showPokemon = true;
      this.showAnswer = true;
      //console.log("Pokemon Page Llamado", pokemonId);
      if (selectId === this.pokemon.id) {
        this.message = `Correcto, ${this.pokemon.name}`;
      } else {
        this.message = `Oops era ${this.pokemon.name}`;
      }
    },
    newGame() {
      (this.showPokemon = false),
        (this.showAnswer = false),
        (this.pokemonArr = []),
        (this.pokemon = null);
      this.mixPokemonArray();
    },
  },

  mounted() {
    this.mixPokemonArray();
  },
};
</script>

