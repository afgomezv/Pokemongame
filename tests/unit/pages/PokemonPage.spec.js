import { shallowMount, mount } from "@vue/test-utils";
import PokemonPage from "@/pages/PokemonPage";
import { pokemons } from "../mocks/pokemons.mock";

describe("PokemonPage Component", () => {
  let wrapper;
  //let mixPokemonArraySpy;

  beforeEach(() => {
    //mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, "mixPokemonArray");
    wrapper = shallowMount(PokemonPage);
  });

  test("debe de hacer match con el snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("debe de llamar mixPokemonArray al montar", () => {
    const mixPokemonArraySpy = jest.spyOn(
      PokemonPage.methods,
      "mixPokemonArray"
    );
    wrapper = shallowMount(PokemonPage);
    expect(mixPokemonArraySpy).toHaveBeenCalled();
  });

  test("debe de hacer match con el snapshot cuando carga los pokemons", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
  test("debe de mostrar los componentes de PokemonPicture y PokemonOption", () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    const picture = wrapper.find("pokemon-picture-stub");
    const options = wrapper.find("pokemon-options-stub");

    //PokemonPicture deben de existir
    expect(picture.exists()).toBeTruthy();
    //PokemonOption deben de existir
    expect(options.exists()).toBeTruthy();
    //PokemonPicture attribute pokemon.id === '1
    expect(picture.attributes("pokemonid")).toBe("1");
    //PokemonOption attribute pokemons toBe true
    expect(options.attributes("pokemons")).toBeTruthy();
  });

  test("pruebas con checkAnswer", async () => {
    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        };
      },
    });

    await wrapper.vm.checkAnswer(1);

    expect(wrapper.find("h2").exists()).toBeTruthy();
    expect(wrapper.vm.showPokemon).toBe(true);
    expect(wrapper.find("h2").text()).toBe(`Correcto, ${pokemons[0].name}`);

    await wrapper.vm.checkAnswer(10);

    expect(wrapper.vm.message).toBe(`Oops, era ${pokemons[0].name}`);
  });
});
