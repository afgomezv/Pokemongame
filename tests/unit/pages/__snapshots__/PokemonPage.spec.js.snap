// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe de hacer match con el snapshot 1`] = `
<h1>Espere por favor...</h1>
<!--v-if-->
`;

exports[`PokemonPage Component debe de hacer match con el snapshot cuando carga los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1><!-- TODO: img -->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- TODO: Opciones -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
</div>
<!--v-if-->
`;
